packages/adblocker/src/resources.ts

```diff
@@ -62,57 +84,88 @@ export default class Resources {
     });
   }
 
-  public static parse(data: string, { checksum }: { checksum: string }): Resources {
-    const typeToResource: Map<string, Map<string, string>> = new Map();
-    const trimComments = (str: string) => str.replace(/^\s*#.*$/gm, '');
-    const chunks = data.split('\n\n');
-
-    for (const chunk of chunks) {
-      const resource = trimComments(chunk).trim();
-      if (resource.length !== 0) {
-        const firstNewLine = resource.indexOf('\n');
-        const split = resource.slice(0, firstNewLine).split(/\s+/);
-        const name = split[0];
-        const type = split[1];
-        const body = resource.slice(firstNewLine + 1);
-
-        if (name === undefined || type === undefined || body === undefined) {
-          continue;
+  public static resolveScriptlet(scriptlet: Scriptlet, scriptlets: Scriptlet[]) {
+    // A LIFO stack saves scriptlet objects in order of dependance.
```

Comment:
in the adblocker-filters we had a quite simple recursive function listing all dependencies:
```js
    const allDependencies = new Set();

    const addDeps = (aScriptlet: any) => {
      for (const dep of aScriptlet.dependencies || []) {
        allDependencies.add(dep);
        const bScriptlet = index.get(dep);
        addDeps(bScriptlet);
      }
    };

    addDeps(scriptlet);

    const deps = [...allDependencies].reverse().map((dep) => index.get(dep).fn);
```

what's the benefit of this new approach?

Reply:
I think, it is related to my question if it is a topological sorting. If so, I would recommend to use a standard algorithm.

Looking at the old code (the simpler example), what I could see is that it will go to a infinite loop if there are cycles (could be that our input never had cycles). The other is that it feels a bit arbitrary to do a depth-first search and reverse the results. In simple graphs, this might work, but not in general. And again, if the order of the dependencies does not matter, than any order would work (e.g. reversing would not be needed either). If the order is irrelevant, then the problem would be not topological sorting, but finding transitive dependencies.

Which is more or less what the old code is doing. To handle cycles gracefully, skipping already visited nodes would be a good addition. And it leaves out the reverse (and if the order is relevant, then you need to compute a topological order and the old algorithm is buggy, since it does not do it properly, in my understanding):

```
    const addDeps = (aScriptlet: any) => {
      for (const dep of aScriptlet.dependencies || []) {
        if (!allDependencies(dep)) {
          allDependencies.add(dep);
          const bScriptlet = index.get(dep);
          addDeps(bScriptlet);
      }
    };

    addDeps(scriptlet);

    const deps = [...allDependencies].map((dep) => index.get(dep).fn);
 ```